contributing/code-style-and-quality-general.md

```diff
@@ -0,0 +1,667 @@
+    * <span style="text-decoration:underline;">Don’t:</span> `Preconditions.checkState(value <= threshold, "value must be below " + threshold)`
+    * <span style="text-decoration:underline;">Don’t:</span> `LOG.debug("value is " + value)`
+    * <span style="text-decoration:underline;">Do:</span> `Preconditions.checkState(value <= threshold, "value must be below %s", threshold)`
+    * <span style="text-decoration:underline;">Do:</span> `LOG.debug("value is {}", value)`
+
+
+### Generics
+
+* **No raw types:** Do not use raw types, unless strictly necessary (sometimes necessary for signature matches, arrays).
+* **Suppress warnings for unchecked conversions:** Add annotations to suppress warnings, if they cannot be avoided (such as “unchecked”, or “serial”). Otherwise warnings about generics flood the build and drown relevant warnings.
+
+
+### equals() / hashCode()
+
+* **equals() / hashCode() should be added when they are well defined only.**
+* They should **not be added to enable a simpler assertion in tests** when they are not well defined. Use hamcrest matchers in that case: [https://github.com/junit-team/junit4/wiki/matchers-and-assertthat](https://github.com/junit-team/junit4/wiki/matchers-and-assertthat)
+* A common indicator that the methods are not well defined is when they take a subset of the fields into account (other than fields that are purely auxiliary).
+* When the methods take mutable fields into account, you often have a design issue. The `equals()`/`hashCode()` methods suggest to use the type as a key, but the signatures suggest it is safe to keep mutating the type.
+
+
+### Java Serialization
+
+* **Do not use Java Serialization for anything !!!**
+* **Do not use Java Serialization for anything !!! !!!**
+* **Do not use Java Serialization for anything !!! !!! !!!**
+*  Internal to Flink, Java serialization is used to transport messages and programs through RPC. This is the only case where we use Java serialization. Because of that, some classes need to be serializable (if they are transported via RPC).
+* **Serializable classes must define a Serial Version UID:**
+
+  `private static final long serialVersionUID = 1L;`
+* **The Serial Version UID for new classes should start at 1** and should generally be bumped on every incompatible change to the class according to the Java serialization compatibility definition (i.e: changing the type of a field, or moving the position of a class in the class hierarchy).
+
+
+### Java Reflection
+
+**Avoid using Java’s Reflection API**
+
+* Java’s Reflection API can be a very useful tool in certain cases but in all cases it is a hack and one should research for alternatives. The only cases where Flink should use reflection are
+    * Dynamically loading implementations from another module (like webUI, additional serializers, pluggable query processors).
+    * Extracting types inside the TypeExtractor class. This is fragile enough any should not be done outside the TypeExtractor class.
```

Comment:
"This is fragile enough any should not be done outside the TypeExtractor class." - Not a complete sentence. Any -> And ?